### ISSUE TITLE
Programmatic control of XML Namespace in SOAP response

### ISSUE THREAD

**Issue by IvanFarkas** [DigDes/SoapCore#387]
Currently, I am using the following to set it statically, see below.
Based on conditions I'd like to set the namespace in the SOAP method call.
How do I have programmatic control of XML Namespace in SOAP response, please?

```C#
public static class ServiceNamespace
{
  public const string Empty = "";
  public const string Value = "http://schemas.company.com/api/v1.0/";
}
```

```C#
[ServiceContract(Namespace = ServiceNamespace.Value)]
public interface ISoapService
{
  [OperationContract(Action = ServiceNamespace.Value + nameof(SoapMethod), ReplyAction = "*")]
  [XmlSerializerFormat(SupportFaults = true)]
  [FaultContract(typeof(Status))]
  SoapMethodResponse SoapMethod(SoapMethodRequest request);

  . . . 
}
```

**Comment by kotovaleksandr** [comment]
No, isn't possible. SoapCore like WCF based on attributes on service and model. If you want - you can develop this feature and send PR to me, i accept them if is does not broke current behavior and covered by tests.

**Comment by IvanFarkas** [comment]
I am trying to change the response namespace based on a request property.
Would it be possible to update ```ServiceBodyWriter._serviceNamespace``` within the SOAP method?
Am I on the right track?

```cs
public ServiceBodyWriter(SoapSerializer serializer, OperationDescription operation, object result, Dictionary<string, object> outResults) : base(isBuffered: true)
{
  . . . 
  _serviceNamespace = operation.Contract.Namespace;
  . . . 
}
```

**Comment by kotovaleksandr** [comment]
@IvanFarkas I think again, whats problem you want solve? Whats namespace you want use on WSDL generation? Service namespace is part of service contract, contract doesn't change on runtime. Maybe make sense use multiple services on different ports or service names?

**Comment by IvanFarkas** [comment]
Hi Kotov,

I do understand your concern from a purist prospective and I do agree. The solution is not elegant.

On the other hand as a pragmatic enterprise developer/consultant I need a way to customize the SOAP response.
I am doing an API modernization for one of the largest Cruise ship company.
Some of their SOAP services was developed over a decade ago and even than was not done by following best practices, to be diplomatic.

They do not want to change the response so the vendors, the consumers of these APIs, do not have to recompile their code.
In many cases requires some serious acrobatics to massage the message to the desired format.

One example of this is this pull request and the other is the issue below.

How do I customize MethodResult?<https://github.com/DigDes/SoapCore/issues/409>

The best way probably would be is some kind of a hook or callback to /src/SoapCore/SoapEndpointMiddleware.cs.
The hook or callback should be able to overwrite element name, namespaces, wrappers, etc.
Essentially we need a middleware to your middleware.

What’s your thought on this, please?

Ivan Farkas
Cloud Solution Architect, 3D HD Soft, LLC<https://3dhdsoft.com/>
Miramar, FL, USA
+1 (954) 931-2990
[LinkedIn]<https://www.linkedin.com/in/ivanfarkas/>  [Twitter] <https://twitter.com/ifarkas>   [Blogger] <https://ivan-farkas.com/>   [GitHub] <https://github.com/IvanFarkas>

**Comment by kotovaleksandr** [comment]
@IvanFarkas I have one idea about customize service namespace. I think we can use IServiceOperationTuner extension for SoapCore (https://github.com/DigDes/SoapCore/blob/develop/src/SoapCore/Extensibility/IServiceOperationTuner.cs). Tune method have operation description object argument, which have reference to service description who store service namespace. Tune method calls before service method. If this case suits for you - i can add public setter to Namespace property in ServiceDescription class.

**Comment by IvanFarkas** [comment]
An ID is passed in the request that indicates the alternate namespace has to be used.
Because of that, your suggestion would not work. Thx, Ivan

**Comment by kotovaleksandr** [comment]
You can retrieve them from HttpContext?

**Comment by IvanFarkas** [comment]
I wish I could. Can't read the stream.

![image](https://user-images.githubusercontent.com/11818001/72905692-ee901080-3cfe-11ea-9897-142a43f2a5a6.png)

**Comment by sv01a** [comment]
it's because PipeStream are closing BodyReader after reading(actual for .net core 3.1)

**Comment by sv01a** [comment]
it's because of PipeStream are closing httpContext.Request.BodyReader: https://github.com/DigDes/SoapCore/blob/develop/src/SoapCore/MessageEncoder/PipeStream.cs#L291

**Comment by kotovaleksandr** [comment]
@IvanFarkas If add method TuneResult to IServiceOperationTuner, its can solve problem?

**Comment by IvanFarkas** [comment]
Perhaps. Let's try.

**Comment by kotovaleksandr** [comment]
@IvanFarkas Check on my fork:
https://github.com/kotovaleksandr/SoapCore
Implement TuneResult method and try change service namespace.
